near-sdk/src/json_types/public_key.rs

```diff
@@ -67,17 +80,19 @@ impl Base58PublicKey {
 
 impl From<Base58PublicKey> for Vec<u8> {
     fn from(v: Base58PublicKey) -> Vec<u8> {
-        v.0
+        let mut out = vec![v.curve as u8];
+        out.extend(v.data);
+        out
     }
 }
 
 impl TryFrom<Vec<u8>> for Base58PublicKey {
     type Error = Box<dyn std::error::Error>;
 
     fn try_from(v: Vec<u8>) -> Result<Self, Self::Error> {
-        match v.len() {
-            33 if v[0] == 0 => Ok(Self(v)),
-            65 if v[0] == 1 => Ok(Self(v)),
+        match &*v {
+            [0, other @ ..] => Ok(Self { curve: CurveType::ED25519, data: other.into() }),
+            [1, other @ ..] => Ok(Self { curve: CurveType::SECP256K1, data: other.into() }),
```

Comment:
prob best to remove the first element since you can just use that allocation, as the `into()` here is sneakily re-allocating.

Reply:
nice -- both will copy data, but your suggestion will avoid the extra allocation

Reply:
but yeah make sure somehow that the length of the data is `>0` as my suggestion can panic if it's empty